src/NUnitFramework/framework/Constraints/ConstraintExpression.cs

```diff
@@ -986,5 +986,17 @@ public Constraint Exist
 
         #endregion
 
+        #region AnyOf
+
+        /// <summary>
+        /// Returns a constraint that tests if an item is equal to any of parameters
+        /// </summary>
+        /// <param name="expected">Expected values</param>
+        public Constraint AnyOf(params object[] expected)
```

Comment:
We're breaking new ground. This is only the second constraint to take more than one argument and the first to accept a variable number of arguments, modeled as a params array. That leads to a number of things we should discuss.

1. Should `AnyOf()` be allowed to compile? We can avoid that by specifying the first argument separately as discussed on the issue.

2. Should `AnyOf(42)` be allowed to compile? We can avoid that by specifying the first two arguments separately.

3. Should we force all arguments to the same type by use of a generic? If so, should there be constraints? A generic without no constraints will allow anything, so it's the same as specifying `object`. However, we could limit this in some ways, depending on how we want to guide folks in using the syntax.

Reply:
1 and 2: Either of these would prevent use of  `.AnyOf(expectedList.ToArray())`. Also, haven't seen collection parameters split out like this before in .NET and it's not appealing to me.

3: If we make it generic, it'll be harder to use because C# as of 7.2) doesn't have the "best common type" proposal implemented. It would force you to write `Is.AnyOf<Animal>(new Cat(), new Dog())` rather than `Is.AnyOf(new Cat(), new Dog())`. I'm not seeing a benefit to that right now.

1. and 2. can be covered with an analyzer shipped in the main package. In fact, I'm going to request that we do exactly this: start an absolutely minimal analyzer that does exactly one thing really well (this diagnostic) and put it in the `\analyzers` folder in our main NuGet package. Once  @mikkelbu thinks the NUnit.Analyzers package is at the right point, we can move it there and have our NUnit package take a dependency on NUnit.Analyzers.

Reply:
@jnm2 As I indicated in the issue, I'm happy to lose that capability. I'd even be happy to drop the `params` array argument and have four constructors with 2, 3, 4 and 5 args respectively. Or more if you like. The point is that the expected value isn't intended to be an open-ended collection but a set of possible values, which can easily be limited to some small number. This isn't a collection constraint, but a different variety of equality constraint.

BTW, the use of "split args" used to be very common. We removed it from the `Constraint` constructor only five years ago. In fact, a single `params` constructor has always worked quite well in C# but sometimes interoperates poorly when there are non-params overloads to choose as well.

If you adamantly oppose the "split args" approach, should we give an error when there is no argument specified? When only one is specified? I know you are thinking about ways you would use the syntax to build up other things, but you can do that by constructing the constraint directly and not using the user-oriented syntax.

Regarding (3), your example seems artificial to me. In any case, you can use the constraint directly if you have a real situation like that.

WRT to an analyzer, that's cool but IMO should not hold up this PR.

Reply:
What about when `expectedList` is a method parameter which is specified alongside each test case, like I brought up in the issue?
For that matter, what's so harmful about a "is contained in this collection" constraint? Data-driven integration tests are common.

I see no need to give a runtime error when zero or one arguments are specified. An analyzer warning seems appropriate to me (but I agree it should not hold up the PR).

I'm sorry it seems artificial! I run into it often. How about this then: `Is.AnyOf<int?>(null, 1)`

Reply:
I agree that there are some problems with generic. In particular, it can give rise to error messages that are hard for the user to figure out. OTOH, your example would be a heck of a lot clearer if you had written `Is.Null.Or.EqualTo(1)`. 😄 

Mainly, I'm trying to make the overall point that a given syntactic element does not have to do every possible thing we can think of. The whole reason that syntax is a separate mini-layer within NUnit is so that you can add lots of syntactic elements that call upon common constraints in differing ways. I am seeing this one as simply being an alternative to use of multiple equality constraints when there are 2 to 5 or so possible values.

Reply:
With `Is.Null.Or.EqualTo(1)` vs `Is.AnyOf(null, 1)`, the latter tempts me. But I get your point.

> I am seeing this one as simply being an alternative to use of multiple equality constraints when there are 2 to 5 or so possible values.

What if I have a whitelist of ten types discovered through reflection and I want to assert that a particular type must be among them?

I'm trying to make the overall point that we should not *raise* the complexity of this thing simply to speculatively deny people certain patterns. That seems almost on the level of raising the complexity to enable speculative patterns.

Reply:
OK... so "some small number" rather than "five". My point is that it's illogical to use this in a more general way as we do in collection constraints. I doubt that you'll have a list of 10,000 types that it might be!

I do agree that we should not raise the complexity of __what the user does__ arbitrarily. Raising the complexity of an implementation detail is a different matter, IMO.

I don't think you and I are going to see eye to eye on this. I'm ~~happy~~ willing to go with your view on the question of generics. I do think we should have an error for zero arguments. NUnit already uses runtime errors in lots of places where we hope to add analysis in the future. Users can always disable the analyzer. If we don't do that, then `AnyOf()` will always fail, which means `Not.AnyOf()` will always succeed! Both need to be wrong! I'm OK to allow just one arg, since that's just a degenerate case.

Reply:
@Dreamescaper After @jnm2's mixing of `ToString` and `Description` I realized you had not overridden `ToString`. My guess is that the base class will result in `<anyof <System.Object[]>>` which is not very useful. Best way to deal with this is to write the Syntax tests (see Syntax directory for examples) and then figure out what you want it to be. It needs to be something we can use in a test to distinguish one constraint from another clearly.

Reply:
> I do agree that we should not raise the complexity of what the user does arbitrarily. Raising the complexity of an implementation detail is a different matter, IMO.

Yes. But taking the example which you said was legitimate, specifying the expected values via method parameter, we are forcing complexity onto the user. The user wants to be able to do `Is.AnyOf(myArrayParameter)`, not:

```cs
myArrayParameter.Length == 1
    ? (Constraint)Is.EqualTo(myArrayParameter[0])
    : Is.AnyOf(myArrayParameter[0], myArrayParameter[1], myArrayParameter.Skip(2).ToArray())
```
(Ironically, the Animal/Cat/Dog problem shows up here; the Constraint cast is necessary.)

One last attempt to get on the same page, and then I'm happy to concede the decision: Given your statement that passing an expectedValues parameter is a legitimate pattern, I feel like you would agree that there should then be an easy way to use that parameter with Is.AnyOf.

While zero items is a degenerate case as well as the single case, I'm okay with an error. The nice thing about runtime errors is that we can start strict and loosen if need be.

Sorry about mixing up the description! `2 or 3 or 5` is marginally verbose but fine if we don't want to do `any of 2, 3, 5`.

Reply:
For your example...

```C#
Assert.That(someValue, new AnyOfConstraint(myArrayParameter);
```

It's definitely not as nice, I agree, but not nearly as bad as the example you gave. Straw man? 😜 

Re Description, I mostly want to make sure that we make an informed decision. Having a test will give us a place to review the decision. Thanks for pointing out that `ToString()` needed to be overridden!

Reply:
No, that's fair! Straw man wasn't my intent. I must have been distracted. So if I was in that position, I'd end up writing a temporary `IsAnyOf(params object[])` method and start an enhancement request. 😊

Reply:
It __is__ pretty interesting that the existence of the fluent syntax gives rise to a desire to be able to use it for everything! It may even be that most people don't realize that the second arg to `Assert.That` is simply a constraint, which can be created any way they like.

Reply:
Considering  "is contained in this collection" constraint and usage for arbitrary collections. Not sure if that makes a lot of sense for unit tests, but might be useful for integration testing, where we can't control everything. Something like "get possible values in one place -> verify actual values in another".

Reply:
@CharliePoole what would you suggest as ToString result? Currently all the collection constraints (it is still a collection constraint underneath) show  <.. Type[]>, not actual values